File: selenium-python/pages/wait_handlers.py
"""Owns the waits, and nothing else — the single place a ``WebDriverWait`` is
constructed.

Three tiers are built once per driver and reused for the driver's whole
lifetime: ``short_wait`` (5 s, negative checks and animation polling),
``default_wait`` (10 s, element interaction), ``long_wait`` (15 s, page load
and SPA hydration). Anything needing another budget calls ``custom(seconds)``.

Element interaction lives in :mod:`pages.element_handler`, which takes its
handlers from here. Page objects and step definitions should use one of the
two — never ``WebDriverWait(...)`` directly.
"""

from __future__ import annotations

from typing import Any, Callable, Optional

from selenium.webdriver.remote.webdriver import WebDriver
from selenium.webdriver.support.ui import WebDriverWait

Locator = tuple[str, str]

# Timeout tiers, in seconds.
SHORT_TIMEOUT = 5
DEFAULT_TIMEOUT = 10
LONG_TIMEOUT = 15

# Poll interval applied to every handler — the Selenium default is 0.5 s.
POLL_INTERVAL = 0.1

_CACHE_ATTR = "_ra_wait_handlers"


class WaitHandlers:
    """Tiered, reusable explicit waits bound to one driver."""

    def __init__(self, driver: WebDriver) -> None:
        self.driver = driver
        self.short_wait = self.custom(SHORT_TIMEOUT)
        self.default_wait = self.custom(DEFAULT_TIMEOUT)
        self.long_wait = self.custom(LONG_TIMEOUT)

    @classmethod
    def for_driver(cls, driver: WebDriver) -> "WaitHandlers":
        """Return the handlers cached on ``driver``, building them on first use.

        The cache lives on the driver object, so it is discarded with the
        driver at the end of each scenario — no stale handler can outlive the
        session it wraps.
        """
        cached = getattr(driver, _CACHE_ATTR, None)
        if cached is None:
            cached = cls(driver)
            setattr(driver, _CACHE_ATTR, cached)
        return cached

    def custom(self, timeout: float) -> WebDriverWait:
        """One-off handler for a timeout the three tiers don't cover."""
        return WebDriverWait(self.driver, timeout=timeout, poll_frequency=POLL_INTERVAL)

    def handler(self, handler: Optional[WebDriverWait] = None) -> WebDriverWait:
        """Resolve an optional handler to the default 10 s tier."""
        return handler if handler is not None else self.default_wait

    # ── Polling ───────────────────────────────────────────────────────────

    def await_true(
        self,
        condition: Callable[[WebDriver], Any],
        handler: Optional[WebDriverWait] = None,
    ) -> bool:
        """Poll ``condition`` until it returns truthy; False on timeout.

        Callers assert on the result instead of catching TimeoutException at
        every call site.
        """
        try:
            return bool(self.handler(handler).until(condition))
        except Exception:
            return False

File: selenium-java/src/test/java/com/example/tests/support/WaitHandlers.java
package com.example.tests.support;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.function.Function;

/**
 * Owns the waits, and nothing else — the single place a {@link WebDriverWait}
 * is constructed.
 *
 * <p>Three tiers are built once per driver and reused for the driver's whole
 * lifetime: {@link #shortWait()} (5 s, negative checks and animation polling),
 * {@link #defaultWait()} (10 s, element interaction), {@link #longWait()}
 * (15 s, page load and SPA hydration). Anything needing another budget calls
 * {@link #custom(Duration)}.
 *
 * <p>Element interaction lives in {@link ElementHandler}, which takes its
 * handlers from here. Page objects and step definitions should use one of the
 * two — never {@code new WebDriverWait(...)}.
 */
public final class WaitHandlers {

    /** Timeout tiers. */
    public static final Duration SHORT_TIMEOUT = Duration.ofSeconds(5);
    public static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(10);
    public static final Duration LONG_TIMEOUT = Duration.ofSeconds(15);

    /** Poll interval applied to every handler — the Selenium default is 500 ms. */
    public static final Duration POLL_INTERVAL = Duration.ofMillis(100);

    /**
     * Per-thread cache. Cucumber runs one scenario per thread with its own
     * driver (see {@code DriverHolder}), so the handlers live exactly as long
     * as the driver does.
     */
    private static final ThreadLocal<WaitHandlers> CACHE = new ThreadLocal<>();

    private final WebDriver driver;
    private final WebDriverWait shortWait;
    private final WebDriverWait defaultWait;
    private final WebDriverWait longWait;

    public WaitHandlers(WebDriver driver) {
        this.driver = driver;
        this.shortWait = custom(SHORT_TIMEOUT);
        this.defaultWait = custom(DEFAULT_TIMEOUT);
        this.longWait = custom(LONG_TIMEOUT);
    }

    /**
     * Returns this thread's handlers, rebuilding them if the thread is now
     * bound to a different driver (new scenario).
     */
    public static WaitHandlers forDriver(WebDriver driver) {
        WaitHandlers cached = CACHE.get();
        if (cached == null || cached.driver != driver) {
            cached = new WaitHandlers(driver);
            CACHE.set(cached);
        }
        return cached;
    }

    /** Drops this thread's handlers — called from the Cucumber @After hook. */
    public static void clear() {
        CACHE.remove();
    }

    public WebDriver driver() {
        return driver;
    }

    // ── Handlers ──────────────────────────────────────────────────────────

    public WebDriverWait shortWait() {
        return shortWait;
    }

    public WebDriverWait defaultWait() {
        return defaultWait;
    }

    public WebDriverWait longWait() {
        return longWait;
    }

    /** One-off handler for a timeout the tiers don't cover. */
    public WebDriverWait custom(Duration timeout) {
        WebDriverWait w = new WebDriverWait(driver, timeout);
        w.pollingEvery(POLL_INTERVAL);
        return w;
    }

    // ── Polling ───────────────────────────────────────────────────────────

    /**
     * Polls {@code condition} until it returns true; false on timeout. Callers
     * assert on the result instead of catching TimeoutException at every site.
     */
    public boolean awaitTrue(Function<WebDriver, Boolean> condition, WebDriverWait handler) {
        try {
            return Boolean.TRUE.equals(handler.until(condition));
        } catch (Exception e) {
            return false;
        }
    }

    /** Same, on the default 10 s tier. */
    public boolean awaitTrue(Function<WebDriver, Boolean> condition) {
        return awaitTrue(condition, defaultWait);
    }
}
